Emit warnings when settings have wrong type

jakul opened the pull request:
Fixes https://github.com/tomchristie/django-rest-framework/issues/2177.

When there is a bad setting the user sees the yellow error page if they're in debug mode, or the server error page if they're not.

![image](https://cloud.githubusercontent.com/assets/862794/5667584/c840d606-9760-11e4-959f-c571f722a689.png)

jakul commented:
Any comments @tomchristie ?

tomchristie commented on rest_framework/exceptions.py:
```diff
@@ -145,3 +145,7 @@ def __init__(self, wait=None, detail=None):
+
+
+class RESTFrameworkSettingHasUnexpectedClassWarning(Warning):
+    pass
```
Let's use python standard exceptions or Django built-in exceptions for configuration errors and suchlike.
We'll reserve our exceptions for user facing API exceptions.

tomchristie commented on rest_framework/settings.py:
```diff
@@ -187,6 +190,25 @@ def __getattr__(self, attr):
+            if issubclass(val.__class__, Iterable) \
+                    and (not issubclass(default.__class__, Iterable)
+                         or isinstance(val, six.string_types)):
+                warnings.warn(
```
Let's just error instead of warning.

tomchristie commented on tests/test_settings.py:
```diff
@@ -15,3 +18,40 @@ def test_import_error_message_maintained(self):
+            'DEFAULT_RENDERER_CLASSES': 'rest_framework.renderers.JSONRenderer'
+        })
+
+        with warnings.catch_warnings(record=True) as w:
```
Don't bother testing the category etc. Just ensure that 'DEFAULT_RENDERER_CLASSES': 'rest_framework.renderers.JSONRenderer' fails, but 'DEFAULT_RENDERER_CLASSES': ['rest_framework.renderers.JSONRenderer'] does not.

jakul replied:
I'm not really sure what you mean by this comment. Should I be adding another test of some sort, or changing this one?

tomchristie commented on rest_framework/settings.py:
```diff
@@ -187,6 +190,25 @@ def __getattr__(self, attr):
+                    and (not issubclass(default.__class__, Iterable)
+                         or isinstance(val, six.string_types)):
+                warnings.warn(
+                    "The `{attr}` setting must be iterable".format(**locals()),
```
Don't use the '**locals()' style. Also more verbose might be good...

    'The "{settings_key}" setting must be a list or tuple, but got {type_name} with value "{value}".'

tomchristie commented:
Seems sensible - reasonable enough for us to just check for string vs list/tuple cases. Note that `None` might be an acceptable value for string cases too.

tomchristie commented:
Bit unclear from our codebase *when* `None` is or isn't acceptable.

jakul commented:
I've made a few updates:

*  changed the text of the error message 
* switched to exceptions instead of warnings

I'm not certain what the next steps are - shall I do an audit and try to work out which settings can be `None` and whitelist them in the type checker?

tomchristie commented on rest_framework/settings.py:
```diff
@@ -187,6 +189,29 @@ def __getattr__(self, attr):
+        else:
+            # Verify that the user hasn't accidentally given a string instead
+            #  of an iterable, or vice-versa
+            default = self.defaults[attr]
```
We can use `val` here.

tomchristie replied:
Oh sorry, ignore that!

tomchristie commented:
> I'm not certain what the next steps are - shall I do an audit and try to work out which settings can be None and whitelist them in the type checker?

Also not sure - perhaps the IMPORT_STRINGS dict should instead be a more general purpose object that includes a list of allowable types (that can be IMPORT_STRING or IMPORT_STRING_LIST.

So eg.

{
    'DEFAULT_AUTHENTICATION_CLASSES': [IMPORT_STRING_LIST],
    'FORMAT_KWARG', [str, NoneType]
    ...
}

Perhaps?

tomchristie commented:
Dropping this as not strictly needed given that we've not found an entirely graceful approach. Don't consider this super neccessary so happy to just drop the ticket unless it sees future interest.
